### ISSUE TITLE
chat: don't delete text if changing tags

### ISSUE THREAD

**Issue by haraldschilly** [sagemathinc/cocalc#6016]
Prerequisite: there is already a tag `#bug`  in a chatroom

## Scenario A

1. In "text mode", type in `foo #bug bar`
2. click on `#bug`

Problem: `foo` is removed, tag is kept. Ok.

Expected: my text is not removed, i.e. it stays `foo #bug  bar`

## Scenario B

1. While `#bug` is still activated from above
2. Type in "text mode": `foo #bug bar`
3. Click on `#bug` tag to deactivate it's filter search

Problem: What remains is `bar`

Expected: Keep `foo #bug bar`, i.e. I just want to see all messages, not get rid of the thread.

## Scenario C

1. No tag active
2. Text: `foo #blah baz` (i.e. a tag besides `#bug`)
3. click on `#bug`

Problem: `foo ` is removed, tag modified.

Expected: Keep my text, just prefix `#bug ` to the entire message.

## Scenario D

1. There are tags `#bug` and `#threads`, not selected, they appear in the same message.
2. Type in text: `foo`
3. Click on `#bug` and then `#blah`, they're prefixed to the message. OK.

Problem: tag `#bug` is no longer in the list of selected tags. It should, because that's how the message starts, right?

4. Click on `#threads` → `#bug` appears as being selected again

Problem: Here, I would expect `#threads` to be removed from my message. At least from the above I expect that deselecting a tag while authoring a message causes tags to be removed.

5. Click on `#bug` to deselect it as well → all tags in my message are removed

Probem: If I do this quickly enough (i.e. not every time I try to add two and remove two tags) I end up with my `foo` text in a block of code (which I can't even remove). Problem seems to be that extra spaces are left and that's interpreted as indented for code. Not being able to remove a block of code (after deleting all characters and spaces from inside of it) seems to be another bug, though.

---

Bottom line, the only valid cases where I would mangle the message is the sequence of the very first tags in the message (to add/remove them). Don't mess up tags inside my message and don't remove regular words. Maybe it's also ok to modify tags at the very end of the message as well, but they must not be in a sentence, i.e. even a "." at the end would guard those tags from being modified.

**Comment by williamstein** [comment]
> Bottom line, the only valid cases where I would mangle the message 

I'm going to get rid of mangling in *all* cases.  We have threads now, so no need to use tags on every message.  
What we do need is a mode where filtering by tag filters the threads (not the messages in threads), so you can see the other messages in a thread.

In any case, the whole point of this tag mangling was to implement threads using tags.  It's not needed, because we have threads now.
